### DISCUSSION TITLE
Constantly override background_image_hsb.hue ?

### DISCUSSION THREAD

**Discussion by Taky0nDon** [wezterm/wezterm#4742] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Operating System: Debian GNU/Linux 12
KDE Plasma Version: 5.27.5
KDE Frameworks Version: 5.103.0
Qt Version: 5.15.8
Kernel Version: 6.1.0-16-amd64 (64-bit)
Graphics Platform: Wayland

### WezTerm version

wezterm 20231221-214626-84ae00c8

### Ask your question!

I wanted to see if I could get the hue of my background image to oscillate from 0 to 3 in 0.1 increments. At first I tried to write a function I just called at the bottom of my config, but I couldn't figure out how to do it without creating an infinite loop. Then I thought it might be cool if I bound it to key presses, so everytime I hit a key, the hue would change by 0.1. I used wezterm.on 

```
`wezterm.on('RawKeyEvent', function (window, pane)
    badcode
end)
```

I'm not sure if RawKeyEvent is the right event name, but it was the closest thing I could find referenced in the documentation. It doesn't do anything.

I did manage to get a janky proof of concept working using the `window-focus-changed` event. Is there an equivalent predefined event for detecting any keypress?

**Comment by wez** [comment, marked as answer]
I'm not sure what your goal is here.  Is it:

* To vary the hue over time?  If so, consider looking at https://wezfurlong.org/wezterm/config/lua/wezterm.time/call_after.html to set up a timer
* To vary the hue when you press a specific key? If so, look at https://wezfurlong.org/wezterm/config/lua/wezterm/action_callback.html
* To vary the hue whenever any key is pressed, without affecting the key press? wezterm doesn't have a way to do this.

If you just want an animated background image, pre-compute the image(s) offline and save/compose them as either a `.gif` or animated `.png` file and assign that as the background, and wezterm will be able to much more efficiently render that into the window that would be possible via lua scripting.

Note that you can control the speed of playback as well (https://wezfurlong.org/wezterm/config/lua/config/background.html#source-definition)

**Reply by Taky0nDon** [reply]
> To vary the hue whenever any key is pressed, without affecting the key press? wezterm doesn't have a way to do this.

This was my immediate goal, although it was really a fall back when I couldn't figure out how to do it passively. I will look into `call_after` and making my own gifs. Thanks for the thorough response!
